// controller/NCBIAPIModule.js
var xml2js = require("xml2js");
var limit = require("simple-rate-limiter");
var request = require("request");
var NcbiApi = require('ncbi-entrez-api');
var RareDisease_Year = require("../model/RareDisease_Year.js");
var MostCitedPublication = require("../model/MostCitedPublication.js");

var callNCBIAPI = limit(
    function(NCBIAPIFunction, args)
    {
        NCBIAPIFunction.apply(null, args);
    })
    .to(3).per(1000).evenly(true);

var getIdsFromSearch = function(searchString, orphanetID, callback)
{
    var pubmed = NcbiApi.dbFactory(NcbiApi.DB.PUBMED);
    pubmed.esearch(
        {
            term: searchString+" AND"+" hasabstract[text]"+" AND"+" Humans[Mesh]",
            sort: "relevance",
            retmode: "json",
            retmax:100000
        },
        function(err, response, data)
        {
            if(err)
            {
                throw err;
            }
            if(data != undefined && data.esearchresult != undefined)
            {
                let idlist=data.esearchresult.idlist;
                if(data.esearchresult.warninglist !== undefined)
                {
                    if(data.esearchresult.warninglist.quotedphrasesnotfound.length !== 0)
                    {
                        idlist = [];
                    }
                }
                console.log("Response from getIdsFromSearch, searchString="+searchString+", orphanetID="+orphanetID+" results: "+idlist.length+" ids");
                callback(idlist, orphanetID);
            }
        }
    );
};

var getRareDisease_YearFromIds = function(ids, orphanetID, callback)
{
    var pubmed = NcbiApi.dbFactory(NcbiApi.DB.PUBMED);
    pubmed.esummary(
        {
            ids: ids,
            retmode: "json"
        },
        function(err, response, data)
        {
            if(err)
            {
                callback(undefined);
            }

            console.log("Response from getRareDisease_YearFromIds, id="+id+", orphanetID="+orphanetID);
            if(data != undefined && data.result != undefined)
            {
                var idslist= data.result.uids;
                var rareDisease_YearList=[];

                for(var i = 0; i < idslist.length; i++)
                {
                    var yearString=data.result[idslist[i]].pubdate.substring(0,4);
                    if(!isNaN(yearString))
                    {
                        var rareDisease_YearIndex=rareDisease_YearList.findIndex(function(currentValue, index, arr){return currentValue.year == yearString});

                        //Is this date already in the array?
                        if(rareDisease_YearIndex != -1)
                        {
                            rareDisease_YearList[rareDisease_YearIndex].numberOfPublications++;
                        }
                        else
                        {
                            //Create rareDisease_Year object
                            var rareDisease_Year=Object.create(RareDisease_Year);
                            rareDisease_Year.init(
                                yearString,
                                orphanetID,
                                1
                            );
                            //Ajout à la collection
                            rareDisease_YearList.push(rareDisease_Year);
                        }
                    }
                }
                callback(rareDisease_YearList);
            }
        }
    );
};

//idList is 200 for max length
var getNumberPublicationsThatCiteGivingPublications = function(idList, callback)
{
    //https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&linkname=pubmed_pubmed_citedin&id=24091329&retmode=json

    var url="https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?id="+idList[0];

    for(let i =1; i < idList.length; i++)
    {
        url += "&id="+idList[i];
    }

    request(
        {
            method: 'POST',
            url: url,
            json: true,
            form:{dbfrom:"pubmed", linkname:"pubmed_pubmed_citedin", retmode:"json"}
        },
        function(err, response, data)
        {
            if(err)
            {
                callback(undefined);
            }
            console.log("Response from getNumberPublicationsThatCiteGivingPublications");
            if(data != undefined && data.linksets != undefined)
            {
                var lengthList=[];
                for(let i = 0; i < data.linksets.length; i++)
                {
                    if(data.linksets[i].linksetdbs != undefined)
                    {
                        lengthList.push(data.linksets[i].linksetdbs[0].links.length);
                    }
                    else
                    {
                        lengthList.push(0);
                    }
                }
                callback(lengthList);
            }
        }
    );
};

var getPublication = function(id, callback)
{
    var pubmed = NcbiApi.dbFactory(NcbiApi.DB.PUBMED);
    pubmed.efetch(
        {
            db: "pubmed",
            retmode: "xml",
            id: id
        },
        function(err, response, data)
        {
            if(err)
            {
                throw(err);
            }
            console.log("Response from getPublication, id="+id);

            //Conversion vers JS Object dans objectFromXml
            var objectFromXml="";
            var parser = new xml2js.Parser();
            parser.parseString
            (
                data,
                function (err, result)
                {
                    objectFromXml = result;
                }
            );

            if(objectFromXml == undefined || objectFromXml.PubmedArticleSet == undefined){return;}

            //Article Title search
            var title=objectFromXml.PubmedArticleSet.PubmedArticle[0].MedlineCitation[0].Article[0].ArticleTitle[0];

            //Abstract Search
            var abstract="";
            var objectAbstract=[];
            try
            {
                objectAbstract=objectFromXml.PubmedArticleSet.PubmedArticle[0].MedlineCitation[0].Article[0].Abstract[0].AbstractText;
            }
            catch (error){}
            finally
            {
                if(objectAbstract.length==1)
                {
                    if(objectAbstract[0].$ != undefined && objectAbstract[0].$ != null)
                    {
                        abstract+=objectAbstract[0].$.Label+" \n";
                        abstract+=objectAbstract[0]._+" \n";
                    }
                    else
                    {
                        abstract=objectAbstract[0];
                    }
                }
                else
                {
                    for(var j=0; j< objectAbstract.length; j++)
                    {
                        if(objectAbstract[j].$ != undefined)
                        {
                            abstract+=objectAbstract[j].$.Label+" \n";
                            abstract+=objectAbstract[j]._+" \n";
                        }
                        else
                        {
                            abstract+=objectAbstract[j]+" \n";
                        }
                    }
                }

                //Authors Search
                var authors=[];
                var objectAuthors=[];
                try
                {
                    objectAuthors=objectFromXml.PubmedArticleSet.PubmedArticle[0].MedlineCitation[0].Article[0].AuthorList[0].Author;
                }
                catch (error){

                }
                finally
                {
                    for(var j=0; j< objectAuthors.length; j++)
                    {
                        var author="";
                        if(objectAuthors[j].LastName != undefined && objectAuthors[j].ForeName != undefined)
                        {
                            author=objectAuthors[j].LastName[0]+" "+objectAuthors[j].ForeName[0];
                        }
                        else if(objectAuthors[j].CollectiveName != undefined)
                        {
                            author=objectAuthors[j].CollectiveName[0];
                        }
                        authors.push(author);
                    }

                    //DOI Search
                    var doi="";
                    var objectDOI=objectFromXml.PubmedArticleSet.PubmedArticle[0].PubmedData[0].ArticleIdList[0].ArticleId;
                    for(var j=0; j< objectDOI.length; j++)
                    {
                        var idObj=objectDOI[j];
                        if(idObj.$.IdType==="doi")
                        {
                            doi=idObj._;
                        }
                    }

                    //Date Search
                    var date="";
                    var objectDate=objectFromXml.PubmedArticleSet.PubmedArticle[0].MedlineCitation[0].DateRevised[0];
                    date=new Date(objectDate.Year+"-"+objectDate.Month+"-"+objectDate.Day+"T00:00:00");

                    //Création de l'objet publication
                    var publication=Object.create(MostCitedPublication);
                    publication.init(
                        id,
                        title,
                        abstract,
                        authors,
                        doi,
                        date,
                        ""
                    );
                    callback(publication);
                }
            }
        }
    );
};

var getPublications = function(idList, callback)
{
    var url="https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi";
    var publications=[];

    var idListString = idList[0];
    for(let i =1; i < idList.length; i++)
    {
        idListString += ","+idList[i];
    }

    request(
        {
            headers: {'content-type' : 'application/x-www-form-urlencoded'},
            method: 'POST',
            url: url,
            form: {db:"pubmed", retmode:"xml", id:idListString}
        },
        function(err, response, data)
        {
            if(err)
            {
                console.error(err);
                fs.writeFile('errors.txt', err, function (err) {
                    if (err) throw err;
                    console.log('It\'s saved!');
                });
                callback([]);
            }
            console.log("Response from getPublications: "+idList.length+" publications");

            //Conversion vers JS Object dans objectFromXml
            var objectFromXml="";
            var parser = new xml2js.Parser();
            parser.parseString
            (
                data,
                function (err, result)
                {
                    if(err)
                    {
                        console.log(err);
                    }
                    objectFromXml = result;
                    if(objectFromXml == undefined || objectFromXml.PubmedArticleSet == undefined)
                    {
                        console.log("Invalid response from NCBI API during getPublications() execution");
                        return;
                    }

                    for(var i = 0; i < objectFromXml.PubmedArticleSet.PubmedArticle.length; i++)
                    {
                        //Id search
                        var id=objectFromXml.PubmedArticleSet.PubmedArticle[i].MedlineCitation[0].PMID[0]._;

                        //Article Title search
                        var title=objectFromXml.PubmedArticleSet.PubmedArticle[i].MedlineCitation[0].Article[0].ArticleTitle[0];

                        //Abstract Search
                        var abstract="";
                        var objectAbstract=[];
                        try
                        {
                            objectAbstract=objectFromXml.PubmedArticleSet.PubmedArticle[i].MedlineCitation[0].Article[0].Abstract[0].AbstractText;
                        }
                        catch (error){}
                        finally
                        {
                            if(objectAbstract.length==1)
                            {
                                if(objectAbstract[0].$ != undefined && objectAbstract[0].$ != null)
                                {
                                    abstract+=objectAbstract[0].$.Label+" \n";
                                    abstract+=objectAbstract[0]._+" \n";
                                }
                                else
                                {
                                    abstract=objectAbstract[0];
                                }
                            }
                            else
                            {
                                for(var j=0; j< objectAbstract.length; j++)
                                {
                                    if(objectAbstract[j].$ != undefined)
                                    {
                                        abstract+=objectAbstract[j].$.Label+" \n";
                                        abstract+=objectAbstract[j]._+" \n";
                                    }
                                    else
                                    {
                                        abstract+=objectAbstract[j]+" \n";
                                    }
                                }
                            }

                            //Authors Search
                            var authors=[];
                            var objectAuthors=[];
                            try
                            {
                                objectAuthors=objectFromXml.PubmedArticleSet.PubmedArticle[i].MedlineCitation[0].Article[0].AuthorList[0].Author;
                            }
                            catch (error){

                            }
                            finally
                            {
                                for(var j=0; j< objectAuthors.length; j++)
                                {
                                    var author="";
                                    if(objectAuthors[j].LastName != undefined && objectAuthors[j].ForeName != undefined)
                                    {
                                        author=objectAuthors[j].LastName[0]+" "+objectAuthors[j].ForeName[0];
                                    }
                                    else if(objectAuthors[j].CollectiveName != undefined)
                                    {
                                        author=objectAuthors[j].CollectiveName[0];
                                    }
                                    authors.push(author);
                                }

                                //DOI Search
                                var doi="";
                                var objectDOI=objectFromXml.PubmedArticleSet.PubmedArticle[i].PubmedData[0].ArticleIdList[0].ArticleId;
                                for(var j=0; j< objectDOI.length; j++)
                                {
                                    var idObj=objectDOI[j];
                                    if(idObj.$.IdType==="doi")
                                    {
                                        doi=idObj._;
                                    }
                                }

                                //Date Search
                                var date="";
                                var objectDate=objectFromXml.PubmedArticleSet.PubmedArticle[i].MedlineCitation[0].DateRevised[0];
                                date=new Date(objectDate.Year+"-"+objectDate.Month+"-"+objectDate.Day+"T00:00:00");

                                //Création de l'objet publication
                                var publication=Object.create(MostCitedPublication);
                                publication.init(
                                    id,
                                    title,
                                    abstract,
                                    authors,
                                    doi,
                                    date,
                                    ""
                                );
                                publications.push(publication);
                            }
                        }
                    }
                    callback(publications);
                }
            );
        }
    );
};


exports.getIdsFromSearch=getIdsFromSearch;
exports.getRareDisease_YearFromIds=getRareDisease_YearFromIds;
exports.getNumberPublicationsThatCiteGivingPublications=getNumberPublicationsThatCiteGivingPublications;
exports.getPublication=getPublication;
exports.getPublications=getPublications;
exports.callNCBIAPI=callNCBIAPI;
